sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java

```diff
@@ -0,0 +1,311 @@
+                                firstProperty = instance.invoke(beanProperty.getReadMethod().get(), List.of());
+                                secondProperty = variableDef.invoke(beanProperty.getReadMethod().get(), List.of());
+                            } else {
+                                continue;
+                            }
+                            ExpressionDef newEqualsExpression = propertyTypeDef.isPrimitive() ?
+                                firstProperty.asCondition(" == ", secondProperty)
+                                : firstProperty.asCondition(" == ", secondProperty)
+                                .asConditionOr(firstProperty.isNonNull().asConditionAnd(
+                                    (propertyTypeDef.isArray()) ?
+                                        ClassTypeDef.of(Arrays.class).invokeStatic("equals", TypeDef.BOOLEAN, Arrays.asList(firstProperty, secondProperty))
+                                        : firstProperty.invoke("equals", TypeDef.BOOLEAN, secondProperty)
+                                ));
+                            if (exp == null) {
+                                exp = newEqualsExpression;
+                            } else {
+                                exp = exp.asConditionAnd(newEqualsExpression);
+                            }
+                        }
+                        return Objects.requireNonNullElseGet(exp, ExpressionDef::trueValue).returning();
+                    })
+                );
+            });
+        classDefBuilder.addMethod(method);
+    }
+
+    /*
+    Creates a hashCode method with signature:
+        public static int BeanNameObject.hashCode(BeanName object)
+     */
+    private static void createHashCodeMethod(ClassDef.ClassDefBuilder classDefBuilder, String objectName, List<PropertyElement> properties) {
+        MethodDef method = MethodDef.builder("hashCode")
+            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+            .addParameter("instance", ClassTypeDef.of(objectName))
+            .returns(int.class)
+            .build((self, parameterDef) -> StatementDef.multi(
+                    parameterDef.get(0).asExpression().isNull().asConditionIf(ExpressionDef.constant(0).returning()),
+                    TypeDef.of(int.class).initialize(ExpressionDef.constant(1)).newLocal("hashValue", hashValue -> {
+                        List<StatementDef> hashUpdates = new ArrayList<>();
+                        TypeDef propertyTypeDef;
+                        ExpressionDef propertyGetter;
+                        ExpressionDef propertyHashCalculation;
+
+                        for (PropertyElement beanProperty : properties) {
+                            propertyTypeDef = TypeDef.of(beanProperty.getType());
+                            if (!beanProperty.hasAnnotation(EqualsAndHashCode.Exclude.class) && beanProperty.getReadMethod().isPresent()) {
```

Comment:
You keep repeating this kind of code everywhere, just check the value and continue; next block initialize a variable and get the value